pkg/ottl/ottlfuncs/func_log.go

```diff
@@ -0,0 +1,67 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type LogArguments[K any] struct {
+	Target ottl.Getter[K] `ottlarg:"0"`
+}
+
+func NewLogFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Int", &LogArguments[K]{}, createLogFunction[K])
+}
+
+func createLogFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*LogArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("IntFactory args must be of type *IntArguments[K]")
+	}
+
+	return logFunc(args.Target), nil
+}
+
+func logFunc[K any](target ottl.Getter[K]) ottl.ExprFunc[K] {
+	return func(ctx context.Context, tCtx K) (interface{}, error) {
+		value, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		switch value := value.(type) {
+		case int64:
+			return math.Log((float64)(value)), nil
+		case string:
+			fltValue, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return nil, nil
+			}
+
+			return math.Log(fltValue), nil
+		case float64:
+			return math.Log(value), nil
+		default:
+			return nil, nil
```

Comment:
We have a new concept of typed-getters that we should use here instead:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/586c62e2241cd18e77a47cd894766a3bdb01376a/pkg/ottl/expression.go#L146-L226

In this case it looks like we want a float64, so we could add a new `FloatLikeGetter` which would behave similar to `StringLikeGetter`.  I'd also be ok with this function being stricter and only allowing int64/float64 and not handling string conversion, in which case we could add a `Float64Getter` that would allow floats and ints.

You can do that as part of this PR or in a separate PR, or if you want I can add it in a separate PR.

Reply:
I'm going to leave this out of the first cut since I'm not sure people using untyped languages are going to be sure that 2.3 is different from "2.3" and don't want to throw unsupported type error when we can try to parse it.

Reply:
I need to document it better, but with the improvements we've made to OTTL's error handling it is good for functions to return errors when they encounter unexpected or undesired situations.  By using typed-getters and returning errors we create consistency across all the OTTL functions.  If users want to ignore the errors they can use `error_mode=ignore`, which is the recommended error mode.

Unfortunately I haven't gotten around to making all the typed-getters so we can still get into situations like these where we need a new one.  I opened up https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21896 to add Float Getters that this function will use.  If you don't want to wait for that PR feel free to add those changes to this PR and use `FloatLikeGetter` and I'll close mine.

Reply:
Actually, lets wait for that PR to merge then do a rebase.  It should get merged today.  It was a large lift than I anticipated and I want to keep the commit history clean now.

Reply:
@mterhar please get the latest from main and use `FloatLikeGetter` as the Getter type.

Reply:
Let me know if I missed something about the update but it looks like it's working with the better getter.